app/src/main/java/ar/teamrocket/duelosmeli/ui/singleplayerActivities/views/NewGameActivity.kt

```diff
@@ -8,66 +8,69 @@ import androidx.recyclerview.widget.LinearLayoutManager
 import ar.teamrocket.duelosmeli.ui.MainMenuActivity
 import ar.teamrocket.duelosmeli.R
 import ar.teamrocket.duelosmeli.data.database.Player
-import ar.teamrocket.duelosmeli.data.database.PlayerDao
-import org.koin.android.ext.android.inject
 import ar.teamrocket.duelosmeli.databinding.ActivityNewGameBinding
 import ar.teamrocket.duelosmeli.ui.singleplayerActivities.adapters.PlayersAdapter
+import ar.teamrocket.duelosmeli.ui.singleplayerActivities.viewModels.NewGameViewModel
+import org.koin.androidx.viewmodel.ext.android.viewModel
 
 class NewGameActivity : AppCompatActivity() {
     private lateinit var binding: ActivityNewGameBinding
-
-    private val playerDao: PlayerDao by inject()
-//TODO: Add View Model
+    private val vm: NewGameViewModel by viewModel()
     override fun onCreate(savedInstanceState: Bundle?) {
         super.onCreate(savedInstanceState)
         binding = ActivityNewGameBinding.inflate(layoutInflater)
         setContentView(binding.root)
+        initUI()
+        setObservers()
+        setListener()
+    }
 
-    binding.iHeader.tvTitle.text=getString(R.string.player)
-    binding.iHeader.ivButtonBack.setOnClickListener { onBackPressed() }
+    private fun initUI() {
+        binding.iHeader.tvTitle.text = getString(R.string.player)
+        vm.setupAllPlayers()
+    }
 
+    private fun setListener() {
+        binding.iHeader.ivButtonBack.setOnClickListener { onBackPressed() }
+        binding.btnStartGame.setOnClickListener {
+            handleButtonStartGame()
+        }
+    }
 
-    //Obtengo todos los jugadores guardados
-        val allPlayers = playerDao.getAll()
-        val allPlayersOrderedByName = playerDao.getAllOrderByName()
+    private fun handleButtonStartGame() {
+        when {
+            vm.playerAlreadyExist(binding.etPlayerName.text) -> {
+                Toast.makeText(this, R.string.player_name_already_exists, Toast.LENGTH_LONG)
+                    .show()
+            }
 
-        val newPlayer = Player("",0) // como el id es autogenerado se le pone 0 y room ya sabe que id poner
+            vm.playerNameIsBlank(binding.etPlayerName.text) -> {
+                Toast.makeText(this, R.string.put_your_name, Toast.LENGTH_LONG).show()
+            }
 
-        binding.btnStartGame.setOnClickListener {
-            newPlayer.name = binding.etPlayerName.text.toString().replace(" ", "")
+            else -> {
+                vm.insertNewPlayer(binding.etPlayerName.text)
+            }
+        }
+    }
 
-            var playerNameAlreadyExists = false
-            for (player in allPlayers){
-                if (player.name == newPlayer.name){
-                    playerNameAlreadyExists = true
-                    break
-                }
+    private fun setObservers() {
+        vm.newPlayerInserted.observe(this) {
+            it?.let {
```

Comment:
La linea 59, "it?.let" cumple la misma funcion que poner "if (it != null)"

Reply:
Yes. Solo que para estos casos es mejor el _**let**_. Kotlin de por si es un lenguaje orientado a funciones, entonces se supone que si tenes una función que te resuelve cierta lógica (que en este caso es una condición de comprobación de nulidad en un if), es mejor usar la función que ya te resuelve eso.
Primero porque ya a la vista es más lindo, menos texto xD, te aseguras de que tenes un código mas "limpio", segundo, que podes reducir el encadenamiento de "ifs" que ya sabemos como nos vieron los profes con eso xD y tercero, a pesar de que se puede usar sin necesidad e un _**let**_, podemos usar el _**run**_ para el caso de que quisiesemos ejecutar algo para el caso de que la variable sea nula.

```
var valorNulo = null
var imprimir = "El valor es: ${valorNulo}"
    
valorNulo?.let {
    println("Por acá no va a entrar")
} ?: run {
    println(imprimir)
}
```

Esto imprime "imprimir".

Ponele que es como un "if-else" mas cheto.